man/dbclean.1
.TH DBCLEAN 1 "12 September 1995"
.FB
.SH NAME
dbclean \- clean a \fIFirstBase\fP database and create a map file
.SH SYNOPSIS
.I dbclean
[\fB-b\fP] [\fB-y\fP] [\fB-f\fP]
\fIdatabase\fP [\fIdatabase\fP ...]
.SH DESCRIPTION
.I Dbclean
cleans a \fIFirstBase\fP database, removing all free and fragmented
storage areas in the database, and creates
a new record map for the database.
This cleaning is done on a byte by byte
basis, making any existing indexes for the database unusable.
(But, see \fIdbregen\fP(1) to regenerate all indexes).
.PP
To determine the amount of free and fragmented space in a \fIFirstBase\fP
database, use \fIdbcheck\fP(8).
.PP
The \fB-b\fP flag will run the \fIFirstBase\fP tool in batch mode,
suitable for shell scripts.
This mode produces no output on the terminal, thus can also be used
when running the process in the background.
The \fB-y\fP flag will run the tool at the terminal without pausing to ask
the user if existing files can be destroyed and without displaying
a 'picture' of the requested task.
.PP
In normal operation, cleaning is done only if a map file is lost,
or disk space needs reclaiming. However, \fIFirstBase\fP databases should be
cleaned once they become static or are to be taken off line.
.PP
The \fB-f\fP flag is used to force a \fIdbclean\fP(1) to run. Normally, if
a database is shut down in an unknown state, i.e. becomes "dirty",
dbclean(1) will not work. The idea is to force the use of \fIdbdump\fP(8) and
\fIdbrestor\fP(8). However, by using the \fB-f\fP
option, dbclean will do its best to clean up the database.
.PP
There are a few \fIFirstBase\fP tools that automatically run \fIdbclean\fP(1)
when they have
completed their tasks. These tools are
\fIdbcgen\fP(1),
\fIdbcat\fP(1),
\fIdbjoin\fP(1),
\fIdbload\fP(1),
and \fIdbpack\fP(1).
All of these tools, except \fIdbpack\fP(1), create
a new database, making the cleaning necessary. \fIDbpack\fP(1) disturbs the
order of the existing records, making post cleaning a necessity also.
.PP
At the end of normal processing, \fIdbclean\fP(1)
runs \fIdbregen\fP(1) to recreate
any autoindexes defined for the \fIFirstBase\fP database. The user only
needs to worry
about any indexes that are non-automatic. These will need to be
regenerated (\fIdbregen\fP(1), \fIdbigen\fP(1)).
.PP
Note: since cleaning a database aligns a new data map with the data file,
the index generators, \fIdbregen\fP(1) and \fIdbigen\fP(1) are passed
the \fB-B\fP argument flag to enable record blocking.
.SH FILES
.PD 0
.TP 10
*.map
\fIFirstBase\fP database map file.
.TP 10
dbase
default name of database.
.TP 10
*.cdb
\fIFirstBase\fP database.
.TP 10
*.ddict
\fIFirstBase\fP database dictionary file.
.TP 10
 .DBCLN_??????
Temporary file used to build the new database.
.PD
.SH SEE ALSO
dbcheck(8), dbregen(1), dbcgen(1), dbcat(1), dbjoin(1), dbload(1),
dbpack(1), dbdump(8), dbrestor(8)
.PP
\fIFirstBase User's Guide and Reference Manual\fP
.br
.SH DIAGNOSTICS
Again, \fIdbclean\fP(1) will make any existing non-automatic
indexes for the \fIFirstBase\fP database
unusable. These indexes will need to be regenerated.
.sp 1
Also, \fIdbclean\fP(1) needs to be run in the same directory as the
database (or at least on the same logical device) as temporary storage is
used in the current working directory to build the new database.
.br


ChangeLog
2012-04-14    <john.bolding@epona.horseweb.com>

	* modified libscr/input.c -- read_kcheck -- to avoid a problem when:
	     EDITINPUT is OFF
	     two (2) ESCAPE's were hanging in read_kcheck

2011-09-29    <john.bolding@epona.horseweb.com>

	* added fflush(-1) to dbmacro via libmexpr/mac_file.c

2009-02-25    <john.bolding@epona.horseweb.com>

	* changed src/dbsql/tokenscript to allow for space between # and define.

2006-09-19  John  <john@cerebus.horseweb.com>

	* complete rewrite of libcc/subline.c to NOT write on the passed in s.

2003-03-29  John Bolding  <john@firstbase.com>

        * tweaked libmpars/tokenscript and dbsql/tokenscript
          to accomodate the bison output "# define" instead of "#define".

2001-09-01  John Bolding  <john@firstbase.com>

        * tweaked configure.in to not use termios.h if OS is sco3.2
	* src/libmexpr/mac_func.c
	* src/libmexpr/mac_str.c
	  augmented dbmacro to include
	     crypt(string, salt), returns an encrypted string ala crypt(3)

2001-03-19  John Bolding  <john@firstbase.com>

	* src/libmexpr/mac_sec.c: 
	* src/libmexpr/mac_func.c: 
	* src/libmexpr/mac_file.c: 
	* src/include/macro.h:
	  modified dbmacro functions as follows:
	  - modified mf_creat() to do a chmod on the file if it already exists.
	  - added an mf_mkstemp() that really calls mf_mktemp()
	    (since mf_mktemp really uses mkstemp(2), it closes the fd.
	     a real mf_mkstemp would modify the array passed in for
	     the user to keep. that did not happen this go around)
	  - renamed all the firstbase secure record functions to 
            r_owner() r_group() r_mode() r_chown() r_chgrp() r_chmod()
	  - added an mf_chmod() command

2001-03-18  John Bolding  <john@firstbase.com>

	* src/dbtools/dbmacro.c: fixed path_trans to be null before
	  its used. fixed to only look at extension of file name, .cgi,
	  to determine whether to go cgi mode. this is so that i can
	  test with program names like ndbmacro.cgi.

	* src/libcdb/cgi.c: modified cgi_read_get to FB_ERROR out
	  if the QUERY_STRING env variable is not set.

	* src/libcdb/System.c: seems linux system() call is a bit
	  flaky at times on some versions, so fb_system() has been
	  modified to not use the system() call at all.
	  now it uses execvp. in addition, changed to waitpid()
	  to try and get better response from cgi dbmacro scripts that
	  use the macro call system(). and, now pays attention to
	  status from waitpid() and exits if its strange.

2001-02-23  John Bolding  <john@firstbase.com>

	* src/libscr/screen.c: modified visible bell area to
	  use a ^G if its not defined.
	  this was a terminfo only bug --- termcap was already working.

2001-02-17  John Bolding  <john@firstbase.com>

	* release 9.3.1
	* man pages for section 4 were missing from distribution.
	  (finally noticed by <garak_dracul@hotmail.com>).

2001-02-16  John Bolding  <john@firstbase.com>

	* man.ps target now in man/Makefile

	* fixed use of mktemp and gets system wide

	* src/libinit/setup.c:
	  changed default behaviour. these are now default:
	     MENUPAUSE OFF
	     MAKEINDEX OFF
	     DECIMAL ON

	* quite a few bugs regarding #if clauses and FSYNC/FCNTL
	  these should have been HAVE_FSYNC or HAVE_FCNTL.
	  some were wrong. some were using FSYNC instead of HAVE_FSYNC.
	  this might have caused multiuser problems with records
	  stored across slow devices (NFS, etc).
	  all of these were introduced when firstbase went autoconf.
	
	* src/libcdb/System.c:
	  - added \" to list of characters that cause a complete system() call
	  - make return status better for those that fork and wait

	* a number of little bugs concering the full screen display
	  for devices bigger than 24 x 80.
	
	* 180 degree about face regarding TERMCAP vs TERMINFO:
	  since ncurses seems to be pretty common on Linux/FreeBSD/Etc,
	  the default is now to use the ncurses library and if this
	  is not found the HAVE_LIBNCURSES is undef'ed in
	  src/include/config.h.

	  Although I am not a TERMINFO fan, perhaps this will be good.
	  At some point, the firstbase screen libraries could
	  make better use of ncurses to optomize screen motion, etc.
	  Perhaps even support for colors, pop-up style ncurses items, etc.

	  So, TERMINFO via ncurses is the default.
	  And, TERMCAP is done only if necessary.

2001-02-05  John Bolding  <john@firstbase.com>

	* src/libscr/input.c: 
	* src/libscr/e_input.c: both of these were modified to handle
	  windows that change sizes while being used, ala SIGWINCH.

	* src/libscr/refresh.c: 
	* src/libscr/scrhdr.c: 
	* src/libscr/scrstat.c: 
	* src/libscr/scrtime.c: 
	* src/libinit/ender.c: 
	* src/libcdb/screrr.c:
	  all of these were modified to better handle standard
	  screen messages on surfaces bigger than 24 x 80,
	  and to restore headers and footers when the frame size changes.

	* src/libcdb/winsize.c: modified to handle different frame/window
	  sizes using an ioctl() call with TIOCGWINSZ or TIOCGSIZE,
	  then allowing environment variables LINES and COLUMNS to
	  override all else. (Thanks to <vanonim@dial.eunet.ch>).

	  finally, sanity checks are done to make sure
	  nothing interactive less than 24 x 80 gets through.

	* src/libinit/getargs.c: took out code that was getting in the way
	  of the new fb_winsize(). all sanity checks are now done at
	  screen setup time.

	* src/libscr/screen.c: force all columns and lines determinations
	  to be done inside of winsize, newly named fb_winsize().

2001-02-02  John Bolding  <john@firstbase.com>

	* src/libinit/setup.c: removed the LINES setup variable,
	     (also force makeindex to be OFF at startup).

2001-02-02  John Bolding  <john@firstbase.com>

	* .visual: changed to /usr/bin/vi (was /usr/ucb/vi).

	* configure.in (INCLUDES): made `ed' and one of `flex' or `lex'
	  and one of `bison', `byacc', or `yacc' required in order
	  to complete ./configure.

$Id: ChangeLog,v 9.9 2006/09/19 17:46:07 john Exp $


src/dbtools/dbclean.c
/*
 * Copyright (C) 2001 FirstBase Software, Inc. All rights reserved.
 *
 * $Id: dbclean.c,v 9.2 2002/07/24 22:01:55 john Exp $
 *
 * See the file LICENSE for conditions of use and distribution.
 *
 */

#ifndef LINT
static char Dbclean_sid[] = "@(#) $Id: dbclean.c,v 9.2 2002/07/24 22:01:55 john Exp $";
#endif

#define FREADWRITE 	(00444|00222)		/* read/write by all */

/*
 *  dbclean - is used to regenerate a fb_database map and rebuild
 *            the fb_database with no fb_free list gaps nor fragmented space.
 *            as long as a SEQF and some kind of header exist
 *            dbclean can recover a dbase, restoring reccnt, delcnt.
 */
 
#include <fb.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fb_vars.h>

#if !FB_PROTOTYPES
static void sclean();
static void makemap();
static void skipbyte();
static void goodrec();
static int nextread();
static int nextwrite();
static int wflush();
static void nextfp();
/*static void usage();*/
#else /* FB_PROTOTYPES */
extern main(int, char **);
static void sclean(char *dname);
static void makemap(int, int, int);
static void skipbyte(int);
static void goodrec(int, int, int, int);
static int nextread(int, char *);
static int nextwrite(int, char *);
static int wflush(int);
static void nextfp(void);
/*static void usage(void);*/
#endif /* FB_PROTOTYPES */

extern short int cdb_secure;
extern long cdb_headsize;
extern char cdb_FREEC;
extern char cdb_FILLC;
extern char cdb_EOREC;
extern char cdb_FILLE;
extern char *cdb_S_FILLC;
extern char *cdb_S_EOREC;
extern char *cdb_S_FILLE;

#define BLKSIZE	10240

char buffer[BLKSIZE + 1];		/* for buffered read */
char wbuffer[BLKSIZE + 1];		/* for buffered write */
int curp = -1;				/* current pointer into buffer */
int maxp = 0;				/* maximum file pointer */
int extrap = 0;				/* extra bytes at block boundry */
int truefields = 0;			/* for counting true # of fields */
long min_rlen = FB_MAPREC_SIZE + 2;
int w_curp = 0;				/* current pointer into wbuffer */

char tempfile[] = ".DBCLN_XXXXXX";	/* file for new dbase (newfd) */
char recnum[20];			/* temp space */
char tbuf[FB_MAXLINE];			/* temporary buffer */
long rpos;				/* for tracking new rpos */
long nrec;				/* for tracking record number */
long ndel;				/* for tracking fb_delete count */
fb_database *hp;			/* pointer to current database */
fb_field *fp = NULL;			/* field pointer of current field */
int nfp = 0;				/* true array offset into kp */
extern short int cdb_allow_dirty;	/* kludge to allow dirty dbase */
char ddname[FB_MAXNAME];		/* for -dd named objects */
char startdb[50];

/*
 * main - mainline logic for dbclean
 */

   main(argc, argv)
   int argc;
   char *argv[];
      {
	 char dname[FB_MAXNAME];
	 int j;
	 
         (void) Dbclean_sid;

         fb_getargs(argc, argv, FB_NODB);
         if (cdb_secure)
            strcpy(startdb, "000000000000000000000000000000");
         else
            strcpy(startdb, "00000000000000000000");
	 ddname[0] = NULL;
	 if ((j = fb_testargs(argc, argv, "-dd")) > 0)
	    strcpy(ddname, argv[j + 1]);
	 if (fb_testargs(argc, argv, "-f") > 0)
	    cdb_allow_dirty = 1;
	 for (j = 1; j < argc; j++){
	    if (!(argv[j][0] == '-')){
	       fb_rootname(dname, argv[j]);
	       sclean(dname);
	       }
	    else if (equal(argv[j], "-dd"))
	       j++;
	    }
         fb_ender();
      }
      
/*
 *  sclean - simple clean of a single fb_database argument
 */      

   static void sclean(dname)
      char *dname;
      
      {
         int newfd, j;
	 char msg[FB_MAXLINE], tname[FB_MAXNAME];
	 int headlen;
	 
	 headlen = cdb_headsize - FB_HEADSTART;
	 hp = fb_dballoc();
 	 fb_dbargs(dname, NIL, hp);
         truefields = 0;
	 if (ddname[0] != NULL){
	    sprintf(tname, SYSMSG[S_FMT_2S], ddname, SYSMSG[S_EXT_DDICT]);
	    fb_mkstr(&(hp->ddict), tname);
	    }
         if (fb_getd_dict(hp) == FB_ERROR)
            fb_xerror(FB_BAD_DICT, hp->ddict, NIL);
         hp->fd = fb_mustopen(hp->dbase, READWRITE);
         if (fb_gethead(hp) == FB_ERROR)
	    fb_xerror(FB_READ_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 fb_scrhdr(hp, "Check Files"); fb_infoline();
	 if (!cdb_batchmode && !cdb_yesmode){
	    fb_move(4, 1);
	    fb_printw("%s%s!!!\n%s\n%s\n",
	       "Make sure no one else is using ", hp->dbase,
	       "After cleaning database, all non-auto indexes",
	       "for that database will require regeneration.");
	    sprintf(msg, "Ready to Clean %s (y/<no>) ?", hp->dbase);
	    if (fb_mustbe('y', msg, cdb_t_lines, 1) != FB_AOK)
	       fb_xerror(FB_ABORT_ERROR, SYSMSG[S_NOT_DONE], NIL);
	    }
	 for (j = 0; j < hp->nfields; j++){
	    if (hp->kp[j]->type != FB_FORMULA && hp->kp[j]->type != FB_LINK)
	       truefields++;
	    if (hp->kp[j]->aid != NULL){	/* abuse msg buffer */
	       sprintf(msg, "%s.idict", hp->kp[j]->aid->autoname);
	       unlink(msg);
	       }
	    }
	 truefields++;			/* for the fb_delete fb_field */
	 close(creat(hp->dmap, 0666));
	 hp->mfd = fb_mustopen(hp->dmap, READWRITE);
	 if (fb_bootmap(hp->mfd) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_MAP], hp->dmap);
	 close(mkstemp(tempfile));
	 newfd = fb_mustopen(tempfile, WRITE);
	 fb_putseq(newfd);	/* put a new sequence here - map is new */
	 if (lseek(newfd, FB_HEADSTART, 0) < 0L)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 if (write(newfd, startdb, (unsigned) headlen) != headlen)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 makemap(hp->fd, newfd, hp->mfd);
         if (fb_puthead(newfd, nrec, ndel) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
         if (cdb_secure)
            if (fb_putmode(newfd, fb_getuid(), fb_getgid(), "666") == FB_ERROR)
	       fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_HEADER], hp->dbase);
	 if (unlink(hp->dbase) < 0)
	    fb_xerror(FB_EXEC_FAILURE, "Could not unlink: ", hp->dbase);
	 if (link(tempfile, hp->dbase) < 0)
	    if (fb_copyfile(tempfile, hp->dbase) < 0)
	       fb_xerror(FB_EXEC_FAILURE, "Could not link/copy: ", tempfile);
         chmod(hp->dbase, FREADWRITE);
	 if (unlink(tempfile) < 0)
	    fb_xerror(FB_EXEC_FAILURE, "Could not unlink: ", tempfile);
	 if (ddname[0] == NULL)
	    sprintf(msg, "dbregen -d %s -b -B", hp->dbase);
	 else
	    sprintf(msg, "dbregen -dd %s -d %s -b -B", ddname, hp->dbase);
	 fb_closedb(hp);
	 fb_refresh();
	 fb_scrstat(SYSMSG[S_AUTOREGEN]); fb_refresh();
	 fb_system(msg, FB_WITHROOT);
      }

/*
 *  makemap - read the fb_database (fd) byte-by-byte, outputting
 *            good records to the new fb_database (nfd) and recording
 *            their spots in the fb_database map (mfd).
 */
 
   static void makemap(fd, newfd, mfd)
      int fd, newfd, mfd;
       
      {
         char c;
	 int n, freeval;
	 
	 nrec = ndel = 0L;
	 if (!cdb_batchmode){ 
	    fb_move(4, 1); fb_clrtobot();
	    fb_scrstat("Cleaning");
	    FB_XWAIT();
	    fb_gcounter(0L);
	    fb_infoline();
	    }
	 if (cdb_yesmode)
	    sleep(1);				/* some are too fast */
	 if (lseek(fd, (long) cdb_headsize, 0) < 0L)
	    fb_xerror(FB_SEEK_ERROR, hp->dbase, NIL);
	 rpos = (long) cdb_headsize;
	 freeval = FB_SLONG*2;
	 for (;;){
	    if ((n = nextread(fd, &c)) == 0)
	       break;
	    else if (n != 1)
	       fb_xerror(FB_READ_ERROR, hp->dbase, NIL);
	    switch(c){
	       case '\006':			/* cdb_FREEC - freelist area */
	          curp += freeval;
	          if (curp > maxp)
		     extrap = curp - maxp;
		  break;
	       case '\030':			/* cdb_FILLC - skip region */
	          skipbyte(fd); break;
	       case '\005':			/* cdb_EOREC - no - op */
	       case '\031':			/* cdb_FILLE - no - op */
	          break;
	       default:				/* must be good */
	          ++nrec;
		  if (!cdb_batchmode){
	             fb_gcounter(nrec);
		     }
	          goodrec((int) c, fd, newfd, mfd);
		  break;
	       }
	    }
	 wflush(newfd);
	    				/* rpos is new avail */
	 if (fb_putmap(mfd, 0L, rpos, 0L, 0L) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_MAP], NIL);
      }
      
/*
 *  skipbyte - all cdb_FILLC's to cdb_EOREC or FILLE
 */

   static void skipbyte(fd)
      int fd;
      
      {
         int n;
	 char c;
	 
	 /* checking for eorec is paranoia since all fillc's end in fille */
         for(c = NULL ; c != cdb_FILLE && c != cdb_EOREC ; ){
	    if ((n = nextread(fd, &c)) == 0)
	       break;
	    else if (n != 1)
	       fb_xerror(FB_READ_ERROR, hp->dbase, NIL);
	    }
      }

/*
 *  goodrec - process a good record
 */

   static void goodrec(ic, fd, newfd, mfd)
      int ic;
      int fd, newfd, mfd;
      
      {
         long rlen, flen;
	 int n, nfields, wlen, fmsg = 0;
         char c;

         c = (char) ic;
	 rlen = 0;
	 nfields = 0;
	 nfp = 0;
	 flen = 0;
	 nextfp();
	 for (;;){
            while (fp->type == FB_BINARY){	/* skip any FB_BINARY fields */
               for (wlen = fp->size; wlen > 0; wlen--){
                  /* skip binary section of data */
                  if (nextwrite(newfd, (char *) &c) != 1)
                     fb_xerror(FB_WRITE_ERROR, hp->dbase, NIL);
                  if ((n = nextread(fd, (char *) &c)) == 0){
                     fb_serror(FB_MESSAGE, "Warning: Unexpected EOF.", NIL);
                     return;
                     }
                  rlen++;
                  }
               nfields++;
               if (nfields < truefields){
                  nfp++;
                  nextfp();
                  }
               flen = fmsg = 0;
               }
	    if (c == '\000'){		/* NULL == fb_field seperator */
	       nfields++;
	       if (nfields < truefields){
		  nfp++;
		  nextfp();
		  }
	       flen = 0;
	       fmsg = 0;
	       }
	    else if (c == cdb_EOREC)
	       break;
	    if (c != '\000')
	       flen++;
	    if (c != '\000' && fp != NULL && flen > fp->size){
	       if (!fmsg){
		  sprintf(tbuf, "Truncating Field %s of record %ld.", 
		     fp->id, nrec);
		  fb_serror(FB_MESSAGE, tbuf, NIL);
		  fmsg = 1;
		  }
	       }
	    else{
	       if (nextwrite(newfd, (char *) &c) != 1)
		  fb_xerror(FB_WRITE_ERROR, hp->dbase, NIL);
	       rlen++; 
	       if (nfp == hp->nfields && c == CHAR_STAR)
	          ndel++;
	       }
	    if ((n = nextread(fd, (char *) &c)) == 0){
	       fb_serror(FB_MESSAGE, "Warning: Unexpected EOF.", NIL);
	       break;
	       }
	    else if (n != 1)
	       fb_xerror(FB_READ_ERROR, hp->dbase, NIL);
	    }
	 if (nfields < truefields){	/* too few fields -- add some */
	    for (nfields++; nfields < truefields; nfields++, rlen++)
	       if (nextwrite(newfd, "\000") != 1)
		  fb_xerror(FB_WRITE_ERROR, hp->dbase, NIL);
	    /* write fb_delete fb_field -- use * to mark this record deleted */
	    nextwrite(newfd, "*");
	    if (nextwrite(newfd, "\000") != 1)
	       fb_xerror(FB_WRITE_ERROR, hp->dbase, NIL);
	    rlen += 2;
	    sprintf(recnum, "%ld", nrec);
	    fb_serror(FB_MESSAGE, "Too Few Fields. Deleting Record ", recnum);
	    ndel++;
	    }
	 if (nextwrite(newfd, cdb_S_EOREC) != 1)
	    fb_xerror(FB_WRITE_ERROR, hp->dbase, NIL);
	 rlen++;
	 if (rlen < min_rlen){			/* fill hole if needed */
	    while (++rlen < min_rlen)
	       nextwrite(newfd, cdb_S_FILLC);
	    nextwrite(newfd, cdb_S_FILLE);		/* rec[rlen] = fille */
	    }
	 if (fb_putmap(mfd, nrec, 0L, rpos, rlen) == FB_ERROR)
	    fb_xerror(FB_IO_ERROR, SYSMSG[S_BAD_MAP], NIL);
	 rpos += rlen;
      }

/*
 * nextread - semi buffered read - read a block if needed.
 */
   static int nextread(fd, buf)
      int fd;
      char *buf;
      
      {
         if (curp < 0 || curp >= maxp){
	    maxp = read(fd, buffer, (unsigned) BLKSIZE);
	    if (maxp <= 0)
	       return(0);
	    curp = 0 + extrap;
	    extrap = 0;
	    }
	 *buf = buffer[curp++];
         return(1);
      }
 
/*
 * nextwrite - semi buffered write - write a block if needed.
 */
   static nextwrite(fd, buf)
      int fd;
      char *buf;
      
      {
         if (w_curp < 0 || w_curp >= BLKSIZE){
	    if (write(fd, wbuffer, (unsigned) BLKSIZE) != BLKSIZE)
	       return(0);
	    w_curp = 0;
	    }
	 wbuffer[w_curp++] = *buf;
         return(1);
      }

/*
 * wflush - flush anything left in the wbuffer
 */
   static wflush(fd)
      int fd;
      
      {
         if (w_curp > 0){
	    if (write(fd, wbuffer, (unsigned) w_curp) != w_curp)
	       return(0);
	    w_curp = 0;
	    }
         return(1);
      }

/*
 * nextfp - get the next valid fb_field descriptor and assign to fp
 */

   static void nextfp()
      {
	 fp = NULL;
         for (; nfp <= hp->nfields; nfp++){
	    if (hp->kp[nfp]->type != FB_FORMULA &&
                  hp->kp[nfp]->type != FB_LINK){
	       fp = hp->kp[nfp];
	       break;
	       }
	    }
	 if (nfp > hp->nfields)
	    fb_serror(FB_MESSAGE, "Can't set next field.", NIL);
      }

/*
 *  usage - for dbclean
 */
/*
*   static void usage()
*      {
*         fb_xerror(FB_MESSAGE,
*	    "usage: dbclean [-dd ddict] [-b] [-y] dbase [dbase ...]", NIL);
*      }
*/
